File: internal/entities/layout.go
package entities

type LayoutDTO struct {
	ID         int32                `json:"id,omitempty"`
	Background *LayoutComponentDTO  `json:"background,omitempty"`
	ImageURL   string               `json:"image_url,omitempty"`
	DesignID   int32                `json:"design_id,omitempty"`
	Width      int32                `json:"width,omitempty"`
	Height     int32                `json:"height,omitempty"`
	Components []LayoutComponentDTO `json:"components,omitempty"`
	Template   Template             `json:"template,omitempty"`
	Grid       GridDTO              `json:"grid,omitempty"`
}

// Prancheta
type Layout struct {
	ID             int32               `json:"id,omitempty"`
	RequestID      int32               `json:"request_id,omitempty"`
	Background     *LayoutComponent    `json:"background,omitempty"`
	BackgroundList []LayoutComponent   `json:"background_list,omitempty"`
	ImageURL       string              `json:"image_url,omitempty"`
	DesignID       int32               `json:"design_id,omitempty"`
	Width          int32               `json:"width,omitempty"`
	Height         int32               `json:"height,omitempty"`
	Components     []LayoutComponent   `json:"components,omitempty"`
	Elements       []LayoutElement     `json:"elements,omitempty"`
	Template       Template            `json:"template,omitempty"`
	Grid           Grid                `json:"grid,omitempty"`
	Stages         []string            `json:"stages,omitempty"`
	Config         LayoutRequestConfig `json:"config,omitempty"`
	DesignAssets   []DesignAsset       `json:"design_assets,omitempty"`
}

func ListToPrioritiesMap(list []string) map[string]int {
	prioritiesMap := make(map[string]int)
	for i, v := range list {
		compType := StringToComponentType(v)
		prioritiesMap[compType.ToString()] = i
	}
	return prioritiesMap
}

File: web/domain/layout/entities/layout.ts
import { LayoutComponentProps } from "./layout_component";
import { LayoutElement, LayoutElementProps } from "./layout_element";

export type LayoutProps = {
  id: number;
  image_url: string;
  design_id: number;
  width: number;
  height: number;
  total?: number;
  done?: number;
  elements?: Array<LayoutElementProps>;
  components?: Array<LayoutComponentProps>;
  background?: LayoutComponentProps
  template?: {
    distortion: {};
    created_at: string;
  };
  grid?: {
    regions: any;
    SlotsX: number;
    SlotsY: number;
  };
  config?: {
    grid?: {
      regions?: any;
      SlotsX?: number;
      SlotsY?: number;
    };
  };
};

export class Layout {
  private p: LayoutProps;
  private _elements: LayoutElement[];
  constructor(p: LayoutProps, elements: LayoutElement[] = []) {
    this.p = p;
    this._elements = elements;
  }

  static create(p: LayoutProps): Layout {
    const elements: LayoutElement[] = [];
    if (p.elements)
      for (const element of p.elements) {
        elements.push(LayoutElement.create(element));
      }

    if (p.components) {
      for (const c of p.components) {
        c.elements.forEach((e) => {
          elements.forEach((el) => {
            if (el.id === e.id) {
              el.setType(c.type);
            }
          });
        });
      }
    }

    return new Layout(p, elements);
  }

  get components(){
    return this.p.components || []
  }

  get bg(){
    return this.p.background
  }

  get id() {
    return this.p.id;
  }

  get designID() {
    return this.p.design_id;
  }

  get imageURL() {
    return this.p.image_url;
  }

  get totalElements() {
    return this.elements.length;
  }

  get elements() {
    return this._elements;
  }

  get width() {
    return this.p.width;
  }

  get height() {
    return this.p.height;
  }
}
